File: frogment/src/main/java/ch/grze/frogment/activity/AbstractStateAwareFrogmentActivity.java
package ch.grze.frogment.activity;

import ch.grze.frogment.State;

//still in Java due to lack of interop: https://youtrack.jetbrains.com/issue/KT-19415
public abstract class AbstractStateAwareFrogmentActivity<S extends State> extends AbstractFrogmentActivity implements StateAwareFrogmentActivityInterface<S> {
    private StateAwareFrogmentActivityComponent<S> stateAwareFrogmentActivityComponent;

    @Override
    public StateAwareFrogmentActivityComponent<S> getStateAwareFrogmentActivityComponent() {
        return stateAwareFrogmentActivityComponent;
    }

    @Override
    public void setStateAwareFrogmentActivityComponent(StateAwareFrogmentActivityComponent<S> component) {
        stateAwareFrogmentActivityComponent = component;
    }
}

File: frogment/src/main/java/ch/grze/frogment/activity/AbstractStateAwareFrogmentActivity.kt
package ch.grze.frogment.activity

import ch.grze.frogment.State
import ch.grze.frogment.StateAware
import ch.grze.frogment.core.backstack.BackStackChangeListener

abstract class AbstractStateAwareFrogmentActivity<S : State> private constructor(override val stateAwareFrogmentActivityComponent: StateAwareFrogmentActivityComponent<S>) :
        AbstractFrogmentActivity(),
        StateAwareFrogmentActivityInterface<S>,
        BackStackChangeListener by stateAwareFrogmentActivityComponent,
        StateAware<S> by stateAwareFrogmentActivityComponent
{
    constructor() : this(StateAwareFrogmentActivityComponent<S>())

    init {
        stateAwareFrogmentActivityComponent.stateAwareFrogmentActivity = this
    }
}
